webapp/tests/forms/steps/lotse/test_personal_data_steps.py

```diff
@@ -114,8 +117,49 @@ def test_if_single_user_then_show_single_text(self, app):
                                                          num=expected_number_of_users)
         with app.test_request_context(method='GET') as req:
             req.session = SecureCookieSession({_LOTSE_DATA_KEY: create_session_form_data(session_data)})
-            step = LotseStepChooser(endpoint='lotse').get_correct_step(StepSteuernummer.name)
+            step = LotseStepChooser(endpoint='lotse').get_correct_step(StepSteuernummer.name, False, ImmutableMultiDict({}))
             step._pre_handle()
 
         assert expected_steuernummer_exists_label == step.form.steuernummer_exists.kwargs['label']
         assert expected_request_new_tax_number_label == step.form.request_new_tax_number.kwargs['label']
+
+
+class TestStepSteuernummerValidate:
+
+    @pytest.mark.usefixtures("test_request_context")
+    def test_if_erica_returns_invalid_tax_number_then_flash_error(self, app):
+        MockErica.tax_number_is_invalid = True
+        bundesland_abbreviation = 'BY'
+        steuernummer = '19811310010'
+        input_data = {'steuernummer_exists': 'yes', 'bundesland': bundesland_abbreviation, 'steuernummer': steuernummer}
+
+        try:
+            with patch('app.forms.steps.lotse.personal_data.flash') as mock_flash:
+                StepSteuernummer.validate_data(ImmutableMultiDict(input_data), {})
+
+        finally:
+            MockErica.tax_number_is_invalid = False
+
+        mock_flash.assert_called_once_with(_('form.lotse.tax-number.invalid-tax-number-error'), 'warn')
+
+    @pytest.mark.usefixtures("test_request_context")
+    def test_if_valid_number_given_then_flash_no_error(self, app):
+        bundesland_abbreviation = 'BY'
+        steuernummer = '19811310010'
+        input_data = {'steuernummer_exists': 'yes', 'bundesland': bundesland_abbreviation, 'steuernummer': steuernummer}
+
+        with patch('app.forms.steps.lotse.personal_data.flash') as mock_flash:
+            StepSteuernummer.validate_data(ImmutableMultiDict(input_data), {})
+
+        mock_flash.assert_not_called()
+
+    @pytest.mark.usefixtures("test_request_context")
+    def test_if_invalid_number_given_then_flash_error(self, app):
+        bundesland_abbreviation = 'BY'
+        steuernummer = '11111111111'
+        input_data = {'steuernummer_exists': 'yes', 'bundesland': bundesland_abbreviation, 'steuernummer': steuernummer}
+
+        with patch('app.forms.steps.lotse.personal_data.flash') as mock_flash:
+            StepSteuernummer.validate_data(ImmutableMultiDict(input_data), {})
+
+            mock_flash.assert_called_once_with(_('form.lotse.tax-number.invalid-tax-number-error'), 'warn')
```

Comment:
Ahh! Thank you for writing that test! Classic case of documenting test - love it.
So, because of this, I'd like to pose the question: Should we maybe only validate the whole form if there is no error on the field level?